### DISCUSSION TITLE
Integrate keystrokes into Main and/or Mixer window

### DISCUSSION THREAD

**Discussion by Chipwingg** [geoffreybennett/alsa-scarlett-gui#119] [Ideas]
Hello! 

First, I would like to say how much I appreciate this software! It has been a lifesaver for configuring my 3rd gen 4i4 over the past couple years. 

As I use my interface for more than desktop music recording, it would be very nice to have a computer keyboard or external numpad toggle mute/0dB for certain dials in the mixer or main window. 

A specific use case is this: When I'm playing music on spotify at a gathering, I would like to enable/disable a microphone (that is routed through the same mix feeding the speakers). This is a basic example, it would be helpful to bind adjustments (like muting a studio monitor, or toggling a particular mix in your headphones) to assignable keystrokes.

**Comment by geoffreybennett** [comment]
Hi @Chipwingg,

Thanks for your message! It's great to hear how people are using this software!

For your use case, I don't think you need alsa-scarlett-gui at all though. The GUI is just a front-end to the kernel's ALSA controls which can easily be adjusted from the command line:

```bash
# Find the ALSA name of your interface (probably USB or USB_1 for a 3rd Gen 4i4)
$ aplay -l
card 0: USB [Scarlett 4i4 USB], device 0: USB Audio [USB Audio]
  Subdevices: 0/1
  Subdevice #0: subdevice #0

# Find the controls you want
$ amixer -c USB controls|grep 'Mix A Input 01'
numid=40,iface=MIXER,name='Mix A Input 01 Playback Volume'

# Get mixer value
$ amixer -c USB cget 'name=Mix A Input 01 Playback Volume'
numid=40,iface=MIXER,name='Mix A Input 01 Playback Volume'
  ; type=INTEGER,access=rw---R--,values=1,min=0,max=172,step=1
  : values=160
  | dBminmax-min=-80.00dB,max=6.00dB

# Set mixer value (0 = Off, 160 = 0dB, 172 = +6dB)
$ amixer -c USB cset 'name=Mix A Input 01 Playback Volume' 172
numid=40,iface=MIXER,name='Mix A Input 01 Playback Volume'
  ; type=INTEGER,access=rw---R--,values=1,min=0,max=172,step=1
  : values=172
  | dBminmax-min=-80.00dB,max=6.00dB
```

You can then put together a shell script to do the thing (set `DEVICE`, `MIX_L`, `MIX_R`, `Input 0X` as appropriate for your use case):
```bash
$ mkdir -p ~/bin
$ cat > ~/bin/alsa-key-control <<'EOT'
#!/bin/sh

if [ $# -ne 2 ]; then
    echo "Usage: $0 <control> <value>"
    exit 1
fi

DEVICE=USB
MIX_L=A
MIX_R=B

CONTROL=$1
VALUE=$2

function set_mixer {
  amixer -q -c $DEVICE cset "name=$1 Playback Volume" -- $2
}

if [ "$CONTROL" = "Mic" ]; then
  set_mixer "Mix $MIX_L Input 01" $VALUE
  set_mixer "Mix $MIX_R Input 01" $VALUE
elif [ "$CONTROL" = "PCM" ]; then
  set_mixer "Mix $MIX_L Input 03" $VALUE
  set_mixer "Mix $MIX_R Input 04" $VALUE
else
  echo "Invalid control: $CONTROL"
  exit 1
fi
EOT
$ chmod +x ~/bin/alsa-key-control
```
You can then use it like this:
```bash
alsa-key-control Mic 0 # turn off the Mic
alsa-key-control Mic 160 # turn on the Mic
```
Finally, use `xbindkeys` or similar to run that shell script with the right options for the moment:
```bash
$ cat > ~/.xbindkeysrc <<'EOT'
"alsacustom Mic 0"
  KP_0
"alsacustom Mic 160"
  KP_1
"alsacustom PCM 0"
  KP_4
"alsacustom PCM 160"
  KP_5
"alsacustom PCM 172"
  KP_6
EOT
$ xbindkeys -n
```
Now keypad-0, 1, 4, 5, 6 can be used to set the Mic and PCM levels without needing to run any app, and the above is easy to adjust for different mixes, inputs, etc. I think the 4i4 is pretty neat in that you've got up to 4 independent stereo sources of audio (2 analogue, 2 from your PC) which you can mix in any combination to 2 independent stereo outputs.

I hope the above makes sense; I assumed a certain familiarity with Linux CLI stuff, so let me know if you need any assistance. I do note you mention "toggle" which I didn't attempt to solve and you might like to take on as a challenge :).

Regards,
Geoffrey.

**Comment by Chipwingg** [comment]
@geoffreybennett Thank you so much for the complete in-depth answer! 

I have a fair amount of CLI experience, but scripting is something I don't have a ton of practice with. At first, when using your example script, I was getting some weird errors about the function `set_mixer`. I noticed when I changed the shebang to bash, `#!/bin/bash`, it ran fine as-is. I also used an alternate function syntax and it works perfectly with `#!/bin/sh`. 
```
set_mixer()
{
  amixer -q -c $DEVICE cset "name=$1 Playback Volume" -- $2
}
```
Which would you recommend, bash or sh? 

Ah yes, when I first said toggle I had in mind something of a mute/un-mute button on a soundboard. I am working on a solution now. The easy thing would be checking if the mixer volume is not zero. If it isn't, making it zero, otherwise, put it to a preset value. 

If I can figure out storing a variable in a separate file though, I'll be able to save the previous level of the mixer output and restore it to that when "un-muting".
